clip/model.py

```diff
@@ -0,0 +1,447 @@
+# Copyright © 2023 Apple Inc.
+
+import json
+from dataclasses import dataclass
+from functools import reduce
+from pathlib import Path
+from typing import Any, List, Optional, Union
+
+import mlx.core as mx
+import mlx.nn as nn
+from config import CLIPConfig, CLIPTextConfig, CLIPVisionConfig
+from huggingface_hub import hf_hub_download
+from mlx.core import linalg as LA
+from mlx.nn.losses import cross_entropy
+from mlx.utils import tree_flatten
+
+
+def quick_gelu(x: mx.array) -> mx.array:
+    """
+    Applies GELU approximation that is fast but somewhat inaccurate. See: https://github.com/hendrycks/GELUs
+    """
+    return x * mx.sigmoid(1.702 * x)
+
+
+def clip_loss(logits_per_text: mx.array, logits_per_image: mx.array) -> mx.array:
+    [N, _] = logits_per_text.shape
+    [M, _] = logits_per_image.shape
```

Comment:
This feels nitpicky but I think it is slightly more readable and far more common to omit the brackets from this destructuring. Basically, `N, _ = logits_per_text.shape` or alternatively `N = len(logits_per_text)`.

Reply:
I will rewrite to match your suggestion.